### DISCUSSION TITLE
Custom URL Scheme? `warg:`?

### DISCUSSION THREAD

**Discussion by esoterra** [bytecodealliance/SIG-Registries#39] [General]
The Component Model intends to use URLs in components to reference other components. The registry will want to use this ability to refer to components in a registry. Since the process of determining the mapping from `(registry, package, version)` to a given content hash requires downloading and validating portions of our cryptographic structure, you cannot simply fetch a component from a given URL on a server and know that it actually is **that** package according to registry semantics.

So, we need some URL that identifies a component but that we don't actually want to be requested over HTTPS in order to retrieve the component. These URLs could use the HTTPS scheme but implicitly have the semantics of our protocol, or they could use a scheme like `warg:` that identifies them as references into the registry namespace with all that entails.

e.g. something like...
* `warg://bytecodealliance.org/foobar/1.0`
* `https://bytecodealliance.org/foobar/1.0`

**Comment by lann** [comment]
A third option for the string embedded in components might be to use "protocol-relative URLs" (relative references):

`//bytecodealliance.org/foobar/1.0`

In some browser context this might resolve to `https:`, while in a registry context it might resolve to `warg:`.

**Reply by esoterra** [reply]
My personal preference would be
1. warg or something like it
2. protocol-relative URL
3. http

I think the more clear it is that this should not be treated as a simple HTTP request, the less likely it is that people will naively choose to resolve packages by HTTP request and end up with vulnerabilities.

**Comment by lukewagner** [comment]
I like the idea of defining a new URL scheme so that we can precisely nail down the fields of information we want to embed in this URL instead of mangling them into an `http:` URL and then worrying about different stages in the toolchain taking the `http:` URL too literally.  Also `warg:` sounds pretty cool :)

In my earlier work on [component examples](https://docs.google.com/presentation/d/11lY9GBghZJ5nCFrf4MKWVrecQude0xy_buE--tnO9kQ/edit#slide=id.gf7c853adf3_0_107), I defaulted to just mangling this info into `http:` since I didn't want to casually propose registering a *second* new scheme with IANA, but if we build some momentum around this (especially if we end up factoring out a W3C CG proposal for the core registry model), then it does seem like we could make a strong case.

With a protocol-relative URL, I worry about, with the absence of protocol literally present in the component, we'll end up with a bunch of implicit assumptions baked into different places that make it hard to evolve over time (although I can't say specifically what).

**Comment by kulakowski-wasm** [comment]
https://www.rfc-editor.org/rfc/rfc7595.html describes the process for registering URI schemes. In particular, [for a provisional scheme](https://www.rfc-editor.org/rfc/rfc7595.html#section-4), which is explicitly intended to be (among other things) a stepping stone towards permanent registration, including for standards-in-progress like this is.

Besides the technical considerations we want to answer for ourselves anyway (the syntax, the semantics, security properties, etc.), I think we need to know exactly who (or what, it can be an organization or a role within one) is the point of contact and the change controller (presumably the BA). And, as I think you've all discussed, ensuring there's no trademark concerns with `warg`.

A final thought: https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml sure does have a lot of very real schemes that never made it past provisional. So it may be that provisional is the *only* step we take.

**Comment by makoConstruct** [comment]
Shouldn't the URL scheme contain a content hash of the component? This would make components with lots of weird dependencies (most real components) resilient against the repositories going down.

> you cannot simply fetch a component from a given URL on a server and know that it actually is that package according to registry semantics

If you include the hash, then it would become the case that you could simply fetch the component from _any random service who claims to have it_ and know that it really is the right package, no matter who they are.

And components would become fully self-describing.

**Reply by lann** [reply]
There are a couple of different scenarios here:

- Fully resolved ("locked") components: As part of a CD pipeline you would likely run a dependency solver on all your components' transitive dependencies, leaving you with a tree of specific component versions. In this case, **yes, absolutely we would want to refer to hashes of specific releases**.
- Libraries: In most cases you don't want to lock (transitive) dependencies to specific versions because you won't pick up bug fixes. Instead you will likely specify some constraint(s) on versions of those dependencies to be resolved by a dependency solver.
- Hybrid: There may be cases where you want a mix of the above, e.g. a collection of interdependent components that all get released together and depend on specific (locked) versions of each other but also have external (version constrained) dependencies.

We haven't formalized a "locked" representation of components, but off the top of my head that could look something like `warg://bytecodealliance.org/foobar/1.0.1?release=abc123...`

**Reply by makoConstruct** [reply]
I believe it's possible to combine content-addressed hashing (with its trustless distributed storage properties) with support for edits, by having the component's descriptor mention its update rules.

A toy example of an edit rule might be: "If you have an update signed by the original author, then that should be served instead. If the version that the server sends is much older than the version had by other servers that the user might check in with, the user will be told of this, and they shall frown, and you may be removed from some federations."

Another edit rule might be "update voted as correct by the listed auditing authority/community"

(Implementation details: Content-addresses wouldn't refer to the component, but to a descriptor, belonging to this distribution system, which contains the content address of the component as well, but if the component has been updated, the component content does not need to be sent. Instead, the descriptor will be sent along with an proof (no, usually weaker than a proof. An attestation.) that the later version is valid according to the edit rule, and the descriptor and the content of the updated version.)

**Reply by lann** [reply]
I believe this sort of "edit rule" could be supported as an implementation policy by our [log-based registry proposal](https://github.com/bytecodealliance/SIG-Registries/blob/proposal/log-based-registry/proposals/Log-Based-Registry.md). In particular, accepting updates to a component that are part of the same "authority chain" as a release you already have is definitely a use case we have in mind. "Update voted as correct by the community" is an interesting one; it would probably be out of scope for our core design but we have discussed "package audits" as a mechanism that could be used to build that sort of feature.

**Reply by makoConstruct** [reply]
Ah, cool. Then I guess my final question is... is it recognized that it would be useful to be able to have a warg URL to a component that doesn't have a name registered in any package repository (where I might be using my own servers for distributing it, or just developing it locally and not really needing to upload it anywhere), where the component's name is either able to alias registered packages (and the system looks at the content-hash first) or where no name is given?

No big deal about community voting for update recognition. It's good enough - and usually preferable - to just assign a specific audit/abuse-report group.
